geoserver/src/eumetsat/pinning/src/main/java/org/geoserver/eumetsat/pinning/ViewsEvaluator.java

```diff
@@ -0,0 +1,543 @@
+        for (String layer : layers) {
+            pinGeoserverLayer(layer, originalTime, mainTime, addPin);
+        }
+    }
+
+    /**
+     * Sync an existing view with the last updated version of it.
+     *
+     * @param view
+     * @param parsed
+     * @throws SQLException
+     * @throws IOException
+     */
+    public void syncView(ViewRecord view, ParsedView parsed) throws SQLException, IOException {
+        Instant originalTime = view.getTimeOriginal();
+        Instant viewTime = parsed.getTime();
+        if (viewTime.equals(view.getTimeMain())) {
+            logger.log(Level.FINE, "View time has not been modified");
+            redoView(view, parsed);
+        } else if (!isExtendingWindow(originalTime, viewTime)) {
+            logger.log(Level.FINE, "View time has been updated outside the pinning window");
+            redoView(view, parsed);
+        } else {
+            logger.log(Level.FINE, "View time has been updated within the pinning window");
+            extendView(view, parsed);
+        }
+        ViewRecord record = buildView(parsed);
+        updateView(record);
+    }
+
+    /**
+     * Fetch a View from DB, given the identifier
+     *
+     * @param viewId
+     * @return
+     * @throws Exception
+     */
+    public ViewRecord fetchView(Long viewId) throws Exception {
+        logger.log(Level.FINE, "Retrieving view: " + viewId);
+        try (PreparedStatement prepStmt = connection.prepareStatement(GET_VIEW_QUERY)) {
+            prepStmt.setLong(1, viewId);
+            try (ResultSet rs = prepStmt.executeQuery()) {
+                if (rs.next()) {
+
+                    Instant timeOriginal =
+                            rs.getTimestamp("time_original", utcCalendar).toInstant();
+                    Instant timeMain = rs.getTimestamp("time_main", utcCalendar).toInstant();
+                    long id = rs.getLong("view_id");
+
+                    Array layersArray = rs.getArray("layers_list");
+                    String[] layers = (String[]) layersArray.getArray();
+                    List<String> layersList = Arrays.asList(layers);
+
+                    Instant lastUpdated = rs.getTimestamp("last_update", utcCalendar).toInstant();
+
+                    ViewRecord record =
+                            new ViewRecord(id, timeOriginal, timeMain, layersList, lastUpdated);
+                    logger.log(Level.FINER, record);
+                    return record;
+                }
+            }
+        }
+        return null;
+    }
+
+    private void setPinLayer(Instant originalTime, Instant mainTime, MappedLayer layer, boolean add)
+            throws SQLException, IOException {
+
+        boolean mainGreaterThanOriginal = mainTime.isAfter(originalTime);
+        Instant minTime = mainGreaterThanOriginal ? originalTime : mainTime;
+        Instant maxTime = mainGreaterThanOriginal ? mainTime : originalTime;
+        Instant start = layer.getNearest(getLeft(minTime));
```

Comment:
This makes me wonder a little... it's nearest matching the left and right, rather than nearest matching the reference time  and then moving left and right... this has consequences, if there are holes in the data (can happen, satellites have maintenance windows), the window might be expanded a lot further than the original view time +/- pin range. It would be moved anyways if snapping the center time, but was some though put into the consequences of doing it one way rather than the other?

Reply:
Fixed.